File: frontend/src/stores/evidence.ts
import { acceptHMRUpdate, defineStore } from "pinia"
import { Delete, List, Reorder, SaveFigure, Upsert } from "wailsjs/go/app/EvidenceService"
import { QueryIn } from "wailsjs/go/app/QueryService"
import { useDataStore } from "~/stores/data"
import { useWorkspacesStore } from "~/stores/workspaces"
import { comparability } from "~/utils/comparability"
import { cycleHolds, pickValues, pinStatus, PIN_METRICS, type PinKind, type PinStatus, type PinValues } from "~/utils/evidence"
import { buildProvenance } from "~/utils/provenance"
import { newestFirst } from "~/utils/scanOrder"
import { sqlLiteral } from "~/utils/sql"

// The evidence board: findings pinned with where they came from, re-checked
// against the newest snapshot of the code (not whichever is open), each
// status naming what it compared with.

export interface Pin {
    id: string
    workspaceId: string
    position: number
    kind: PinKind
    entityKey: string
    title: string
    route: string
    scanId: string | null
    headCommit: string
    revision: number
    lens: string
    scope: string
    role: string
    values: string
    note: string
    figurePath: string
    createdAt: string
}

export interface PinInput {
    kind: PinKind
    entityKey: string
    title: string
    values?: PinValues
    note?: string
    /** A figure to keep with a view pin, as PNG base64. */
    figure?: string | null
}

export const useEvidenceStore = defineStore("evidence", {
    state: () => ({
        workspace: "" as string,
        pins: [] as Pin[],
        now: {} as Record<string, PinValues | null>,
        statuses: {} as Record<string, PinStatus>,
        checkedAgainst: "" as string,
        loading: false,
    }),
    getters: {
        count: (s) => s.pins.filter(p => p.kind !== "heading").length,
        isPinned: (s) => (kind: PinKind, key: string) => s.pins.some(p => p.kind === kind && p.entityKey === key),
        valuesOf: () => (p: Pin): PinValues => { try { return JSON.parse(p.values || "{}") } catch { return {} } },
    },
    actions: {
        async load(workspaceId: string, force = false) {
            if (this.workspace === workspaceId && !force) return
            this.workspace = workspaceId
            try { this.pins = ((await List(workspaceId)) ?? []) as any } catch { this.pins = [] }
            void this.recheck()
        },
        async pin(input: PinInput): Promise<Pin | null> {
            const ws = useWorkspacesStore()
            const data = useDataStore()
            if (!ws.active) return null
            const p = buildProvenance()
            const route = typeof location !== "undefined" ? location.hash.replace(/^#/, "") : ""
            const saved = (await Upsert({
                id: "", workspaceId: ws.active.id, position: 0, kind: input.kind, entityKey: input.entityKey, title: input.title,
                route, scanId: ws.openScanId ?? null, headCommit: p.commit, revision: data._snapshotRevision ?? 0,
                lens: p.lens ?? "", scope: p.scope ?? "", role: p.role ?? "", values: JSON.stringify(input.values ?? {}),
                note: input.note ?? "", figurePath: "", createdAt: new Date().toISOString(),
            } as any)) as any as Pin
            if (input.figure) {
                try {
                    saved.figurePath = await SaveFigure(ws.active.id, saved.id, input.figure)
                    await Upsert(saved as any)
                } catch { /* the pin stands without its figure */ }
            }
            this.pins = [...this.pins, saved]
            void this.recheck()
            return saved
        },
        async update(id: string, patch: Partial<Pick<Pin, "title" | "note">>) {
            const p = this.pins.find(x => x.id === id)
            if (!p) return
            Object.assign(p, patch)
            await Upsert(p as any)
        },
        async heading(title: string) {
            const ws = useWorkspacesStore()
            if (!ws.active) return
            const saved = (await Upsert({ id: "", workspaceId: ws.active.id, position: 0, kind: "heading", entityKey: "", title, route: "", scanId: null, headCommit: "", revision: 0, lens: "", scope: "", role: "", values: "{}", note: "", figurePath: "", createdAt: new Date().toISOString() } as any)) as any as Pin
            this.pins = [...this.pins, saved]
        },
        async reorder(ids: string[]) {
            const byId = new Map(this.pins.map(p => [p.id, p]))
            this.pins = ids.map(id => byId.get(id)!).filter(Boolean)
            await Reorder(this.workspace, ids)
        },
        async remove(id: string) {
            await Delete(id)
            this.pins = this.pins.filter(p => p.id !== id)
        },

        /** Every pin against the newest complete snapshot. */
        async recheck() {
            const ws = useWorkspacesStore()
            const data = useDataStore()
            const newest: any = newestFirst(ws.scans.filter((s: any) => s.status === "complete"))[0]
            if (!newest) return
            this.checkedAgainst = newest.id
            const q = (sql: string) => QueryIn(newest.id, sql) as Promise<any[]>
            const now: Record<string, PinValues | null> = {}
            const statuses: Record<string, PinStatus> = {}
            let groupOf: Map<string, string> | null = null
            for (const p of this.pins) {
                if (p.kind === "heading" || p.kind === "view") continue
                const pinnedScan: any = p.scanId ? ws.scans.find((s: any) => s.id === p.scanId) : null
                const c = pinnedScan ? comparability(pinnedScan, newest) : { ok: true, reasons: [] as Array<{ level: string; text: string }> }
                const blocked = c.ok ? null : c.reasons.find(r => r.level === "block")?.text ?? "different analyses"
                let values: PinValues | null = null
                try {
                    if (p.kind === "component" || p.kind === "file") {
                        const table = p.kind === "component" ? "components" : "files"
                        const [row] = await q(`SELECT * FROM ${table} WHERE name = ${sqlLiteral(p.entityKey)} LIMIT 1`)
                        values = row ? pickValues(row, PIN_METRICS[p.kind]) : null
                    } else if (p.kind === "cycle") {
                        if (!groupOf) {
                            groupOf = new Map()
                            for (const r of await q(`SELECT "group", component FROM component_strongly_connected_groups WHERE "group" IN (SELECT "group" FROM component_strongly_connected_groups GROUP BY 1 HAVING count(*) > 1)`)) groupOf.set(String(r.component), String(r.group))
                        }
                        const members = p.entityKey.split("\n")
                        values = cycleHolds(members, groupOf) ? { size: members.length } : null
                    } else if (p.kind === "rule") {
                        const [rule, from, to, file] = p.entityKey.split("|")
                        const [r] = await q(`SELECT count(*) AS n FROM rules WHERE status = 'violation' AND rule = ${sqlLiteral(rule)} AND "from" = ${sqlLiteral(from)} AND "to" = ${sqlLiteral(to)} AND file = ${sqlLiteral(file ?? "")}`)
                        values = Number(r?.n) ? { findings: Number(r.n) } : null
                    } else if (p.kind === "pair") {
                        const [from, to] = p.entityKey.split(">")
                        const [r] = await q(`SELECT sum(reference_count) AS n FROM component_connections_direct WHERE "from" = ${sqlLiteral(from)} AND "to" = ${sqlLiteral(to)}`)
                        values = Number(r?.n) ? { references: Number(r.n) } : null
                    }
                } catch { values = null }
                now[p.id] = values
                statuses[p.id] = pinStatus({ pinned: this.valuesOf(p), now: values, blocked, deleted: !p.scanId, label: id => data.statNiceName(id) || id })
            }
            this.now = now
            this.statuses = statuses
        },
    },
})

if (import.meta.hot) import.meta.hot.accept(acceptHMRUpdate(useEvidenceStore, import.meta.hot))

File: app/store/evidence.go
package store

import (
	"database/sql"
	"time"
)

// Pin is one finding on a workspace's evidence board.
type Pin struct {
	ID          string    `json:"id"`
	WorkspaceID string    `json:"workspaceId"`
	Position    int       `json:"position"`
	Kind        string    `json:"kind"`
	EntityKey   string    `json:"entityKey"`
	Title       string    `json:"title"`
	Route       string    `json:"route"`
	ScanID      *string   `json:"scanId"`
	HeadCommit  string    `json:"headCommit"`
	Revision    int       `json:"revision"`
	Lens        string    `json:"lens"`
	Scope       string    `json:"scope"`
	Role        string    `json:"role"`
	Values      string    `json:"values"`
	Note        string    `json:"note"`
	FigurePath  string    `json:"figurePath"`
	CreatedAt   time.Time `json:"createdAt"`
}

const pinColumns = `id, workspace_id, position, kind, entity_key, title, route, scan_id, head_commit, revision, lens, scope, role, pinned_values, note, figure_path, created_at`

func scanPin(scan func(dest ...any) error) (*Pin, error) {
	p := &Pin{}
	var scanID sql.NullString
	if err := scan(&p.ID, &p.WorkspaceID, &p.Position, &p.Kind, &p.EntityKey, &p.Title, &p.Route, &scanID, &p.HeadCommit, &p.Revision, &p.Lens, &p.Scope, &p.Role, &p.Values, &p.Note, &p.FigurePath, &p.CreatedAt); err != nil {
		return nil, err
	}
	if scanID.Valid {
		p.ScanID = &scanID.String
	}
	return p, nil
}

// Pins lists a workspace's board in order.
func (s *Store) Pins(workspaceID string) ([]*Pin, error) {
	rows, err := s.db.Query(`SELECT `+pinColumns+` FROM evidence_pins WHERE workspace_id = ? ORDER BY position, created_at`, workspaceID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []*Pin
	for rows.Next() {
		p, err := scanPin(rows.Scan)
		if err != nil {
			return nil, err
		}
		out = append(out, p)
	}
	return out, rows.Err()
}

// UpsertPin inserts a pin at the end of its board, or updates one in place.
func (s *Store) UpsertPin(p *Pin) error {
	var exists int
	_ = s.db.QueryRow(`SELECT count(*) FROM evidence_pins WHERE id = ?`, p.ID).Scan(&exists)
	var scanID any
	if p.ScanID != nil && *p.ScanID != "" {
		scanID = *p.ScanID
	}
	if exists > 0 {
		_, err := s.db.Exec(`UPDATE evidence_pins SET title = ?, note = ?, route = ?, pinned_values = ?, figure_path = ? WHERE id = ?`, p.Title, p.Note, p.Route, p.Values, p.FigurePath, p.ID)
		return err
	}
	var next int
	_ = s.db.QueryRow(`SELECT coalesce(max(position), -1) + 1 FROM evidence_pins WHERE workspace_id = ?`, p.WorkspaceID).Scan(&next)
	if p.CreatedAt.IsZero() {
		p.CreatedAt = time.Now().UTC()
	}
	_, err := s.db.Exec(`INSERT INTO evidence_pins (`+pinColumns+`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		p.ID, p.WorkspaceID, next, p.Kind, p.EntityKey, p.Title, p.Route, scanID, p.HeadCommit, p.Revision, p.Lens, p.Scope, p.Role, p.Values, p.Note, p.FigurePath, p.CreatedAt)
	return err
}

// ReorderPins sets the board order to the ids given.
func (s *Store) ReorderPins(workspaceID string, ids []string) error {
	tx, err := s.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	for i, id := range ids {
		if _, err := tx.Exec(`UPDATE evidence_pins SET position = ? WHERE id = ? AND workspace_id = ?`, i, id, workspaceID); err != nil {
			return err
		}
	}
	return tx.Commit()
}

// DeletePin removes a pin and returns its figure path, for the caller to delete.
func (s *Store) DeletePin(id string) (string, error) {
	var fig string
	_ = s.db.QueryRow(`SELECT figure_path FROM evidence_pins WHERE id = ?`, id).Scan(&fig)
	_, err := s.db.Exec(`DELETE FROM evidence_pins WHERE id = ?`, id)
	return fig, err
}
